src/core.ts

```diff
@@ -1,133 +1,90 @@
+import { FlattenOperator } from './operator/flatten'
+import { RememberOperator } from './operator/remember'
+
+const EMPTY = {}
+function noop (): void { return void 0 }
+
+class FromPromiseProducer<T> implements Subscribable<T> {
+  private on: boolean = false
+  constructor (private promise: Promise<T>) {}
+
+  subscribe (listener: Listener<T>) {
+    this.on = true
+    this.promise.then(
+      (x: T) => {
+        if (this.on) {
+          listener.next(x)
+          listener.complete()
+        }
+      },
+      (e: any) => {
+        listener.error(e)
+      }
+    ).then(null, (err: any) => {
+      setTimeout(() => { throw err })
+    })
+    return new StreamSubscription(() => { this.on = false })
   }
-  return b;
 }
 
-export interface InternalListener<T> {
-  _n: (v: T) => void;
-  _e: (err: any) => void;
-  _c: () => void;
-}
-
-export const NO_IL: InternalListener<any> = {
-  _n: noop,
-  _e: noop,
-  _c: noop,
-};
-
-export interface InternalProducer<T> {
-  _start: (listener: InternalListener<T>) => void;
-  _stop: () => void;
-}
+class PeriodicProducer implements Subscribable<number> {
+  constructor (private period: number) {}
 
-export interface OutSender<T> {
-  out: Stream<T>;
-}
-
-export interface Operator<T, R> extends InternalProducer<R>, InternalListener<T>, OutSender<R> {
-  type: string;
-  ins: Stream<T>;
-  _start: (out: Stream<R>) => void;
-}
+  subscribe (listener: Listener<number>) {
+    let i: number = -1
+    function intervalHandler () { listener.next(i++) }
+    let intervalID: number = -1
+    intervalID = (<number> <any> setInterval(intervalHandler, this.period))
 
-export interface Aggregator<T, U> extends InternalProducer<U>, OutSender<U> {
-  type: string;
-  insArr: Array<Stream<T>>;
-  _start: (out: Stream<U>) => void;
-}
-
-export interface Producer<T> {
-  start: (listener: Listener<T>) => void;
-  stop: () => void;
-}
-
-export interface Listener<T> {
-  next: (x: T) => void;
-  error: (err: any) => void;
-  complete: () => void;
+    return new StreamSubscription(() => {
+      if (intervalID !== -1) clearInterval(intervalID)
+      intervalID = -1
+      i = -1
+    })
+  }
 }
 
-// mutates the input
-function internalizeProducer<T>(producer: Producer<T>) {
-  (<InternalProducer<T>> (<any> producer))._start =
-    function _start(il: InternalListener<T>) {
-      (<Listener<T>> (<any> il)).next = il._n;
-      (<Listener<T>> (<any> il)).error = il._e;
-      (<Listener<T>> (<any> il)).complete = il._c;
-      this.start(<Listener<T>> (<any> il));
-    };
-  (<InternalProducer<T>> (<any> producer))._stop = producer.stop;
-}
+class MergeProducer<T> implements Subscribable<T> {
+  constructor (private streamsArray: Stream<T>[]) {}
 
-function compose2<T, U>(f1: (t: T) => any, f2: (t: T) => any): (t: T) => any {
-  return function composedFn(arg: T): any {
-    return f1(f2(arg));
-  };
-}
+  subscribe (listener: Listener<T>) {
+    const sa = this.streamsArray
```

Comment:
Just to keep in mind: it's important that merge and combine implement the Aggregator interface because the field `insArr` will be used by devtools to traverse the stream graph.